File: opcua-webservice-node/src/server/service/transactions/base/UaReadNodeTransaction.ts
import { StatusCodes, TimestampsToReturn } from "opcua-webapi";
import { makeUaStatusCode, UaDataValue, UaError, UaVariant } from "opcua-webapi-ts";
import { NodeManager } from "../../../addressspace/nodemanager";
import { ReadContext } from "../../..";
import { UaReadTransaction } from "./UaReadTransaction";

export class UaReadNodeTransaction extends UaReadTransaction {
    private readonly _nodeManager: NodeManager;

    constructor(
        context: ReadContext,
        handleIds: Array<number>,
        nodeManager: NodeManager,
    ) {
        super(context, handleIds);
        this._nodeManager = nodeManager;
    }

    async execute(): Promise<void> {
        try {
            const now = new Date();
            const nodesToRead = this.getRequestedItems();

            for (const item of nodesToRead) {
                const result = this._nodeManager.read(
                    item.nodeId,
                    item.attributeId,
                    now,
                    this.timestampsToReturn as TimestampsToReturn,
                );
                this._results.push(result);
            }
        } catch (error) {
            this.buildErrorResults(error);
        }
    }
}

File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/service/transactions/base/UaReadNodeTransaction.java
package org.opcfoundation.webserver.service.transactions.base;

import org.eclipse.milo.opcua.stack.core.types.builtin.DataValue;
import org.eclipse.milo.opcua.stack.core.types.builtin.DateTime;
import org.eclipse.milo.opcua.stack.core.types.structured.ReadValueId;
import org.opcfoundation.webserver.addressspace.nodemanager.NodeManager;
import org.opcfoundation.webapi.service.types.ReadContext;

import java.util.List;
import java.util.concurrent.CompletableFuture;

public class UaReadNodeTransaction extends UaReadTransaction {
    private final NodeManager nodeManager;

    public UaReadNodeTransaction(
            ReadContext context,
            List<Integer> handleIds,
            NodeManager nodeManager)
    {
        super(context, handleIds);
        this.nodeManager = nodeManager;
    }

    public CompletableFuture<Void> execute()
    {
        return CompletableFuture.supplyAsync(()-> {
            try
            {
                DateTime now = DateTime.now();
                List<ReadValueId> nodesToRead = getRequestedItems();

                for (ReadValueId item: nodesToRead)
                {
                    DataValue result = nodeManager.read(item.getNodeId(), item.getAttributeId().intValue(), now, timestampsToReturn);
                    results.add(result);
                }

                return null;
            } catch (Exception e) {
                buildErrorResponse(e);
            }

            return null;
        });
    }
}
